File: leetcode/jump-game.py
'''
https://leetcode.com/problems/jump-game/
'''


class Solution:                               
    # @param A, a list of integers            
    # @return a boolean                       
    def canJump(self, A):
            length   = len(A)
            #if A is a [] list
            if length == 0:
                    return False
            
            # init maxDis = 0, 
            # because BEFORE ARRIVING at the index 0, we can't jump any distance, that is 0.
            maxDis = 0
            for i in range(length):
                    if maxDis < i:
                            return False

                    if maxDis< A[i] + i:
                            maxDis = A[i]+i
                           
                    if maxDis>= len(A)-1:
                            return True

File: leetcode/jump-game.c
/*
https://leetcode.com/problems/jump-game/



*/

bool canJump(int A[], int n) {
    
    //极端情况
    if (n == 0 || n == 1)
        return true;
    //farest: 在当前下标处，仍能到达的最远距离
    //初始化
    int farest = A[0];
    int faresttmp;
    int i;
    for(i = 1; i<n; i++ ){
        if(farest >= n)
            return true;
        
        //当前位置在最远距离的范围内时，仍能继续
        if(i <= farest){
             faresttmp = i + A[i];
             if(faresttmp > farest)
                farest = faresttmp;
             
        }
        else{
            return false;
        }
    }
}
